src/t8_schemes/t8_new_feature/t8_subelements/t8_subelements_quad_cxx.cxx

```diff
@@ -0,0 +1,1325 @@
+    if (binary_array[k] == 1) {
+      sub_id -= 2; /* two subelements adjacent to face k */
+    }
+    else {
+      sub_id -= 1; /* one subelement adjacent to face k */
+    }
+    if (sub_id == 0) {
+      sub_face_id = 0; /* in this case, the subelement is the first element of the face */
+    }
+    k++;
+  }
+  face_number = k - 1;
+
+  if (binary_array[face_number] == 0) {
+    split = 0; /* the face is not split */ 
+  }
+  else {
+    split = 1; /* the face is split */
+  }
+
+  location[0] = face_number;
+  location[1] = split;
+  location[2] = sub_face_id;
+}
+
+t8_element_shape_t
+t8_default_scheme_sub_c::t8_element_get_shape (const t8_element_t * elem)
+{
+  const t8_quad_with_subelements *pquad_w_sub = (const t8_quad_with_subelements *) elem;
+
+  T8_ASSERT (t8_element_is_valid (elem));
+
+  if (pquad_w_sub->dummy_is_subelement != 1) { 
+    return T8_ECLASS_QUAD;
+  }
+  else { /* for quads, all subelements are triangles */
+    return T8_ECLASS_TRIANGLE;
+  }
+} 
+
+void
+t8_default_scheme_sub_c::t8_element_new (int length, t8_element_t ** elem)
+{
+  /* allocate memory for a quad with subelements */
+  t8_default_scheme_common_c::t8_element_new (length, elem);
+
+  int                 i;
+  for (i = 0; i < length; i++) {
+    t8_quad_with_subelements *pquad_w_sub = (t8_quad_with_subelements *) elem[i];
+    t8_element_init (1, elem[i], 0);
+    /* set dimension of quad to 2 */
+    T8_QUAD_SET_TDIM ((p4est_quadrant_t *) & pquad_w_sub->p4q, 2);
+  }
+}
+
+void
+t8_default_scheme_sub_c::t8_element_init (int length, t8_element_t * elem,
+                                          int new_called)
+{
+  t8_quad_with_subelements *pquad_w_sub = (t8_quad_with_subelements *) elem;
+
+  int                 i;
+
+  for (i = 0; i < length; i++) {
+    /* initalize subelement parameters */
+    pquad_w_sub[i].dummy_is_subelement = 0;
+    pquad_w_sub[i].subelement_type = -1;
+    pquad_w_sub[i].subelement_id = -1;
+    pquad_w_sub[i].num_subelement_ids = -1;
+
+    #ifdef T8_ENABLE_DEBUG
+    /* In debugging mode we iterate over all length many elements and 
+     * set their quad to the level 0 quad with ID 0. */
+    if (!new_called) {
+      p4est_quadrant_t   *quad = &pquad_w_sub[i].p4q;
+      p4est_quadrant_set_morton (quad, 0, 0);
+      T8_QUAD_SET_TDIM (quad, 2);
+      T8_ASSERT (p4est_quadrant_is_extended (quad));
+    }
+  }
+#endif
+}
+
+#ifdef T8_ENABLE_DEBUG
+/* *INDENT-OFF* */
+/* indent bug, indent adds a second "const" modifier */
+int
+t8_default_scheme_sub_c::t8_element_is_valid (const t8_element_t * elem) const 
+/* *INDENT-ON* */
+{
+  const t8_quad_with_subelements *pquad_w_sub = (const t8_quad_with_subelements *) elem;
+  const p4est_quadrant_t *q = &pquad_w_sub->p4q;
+
+  T8_ASSERT (pquad_w_sub->dummy_is_subelement == 0 || pquad_w_sub->dummy_is_subelement == 1);
+  T8_ASSERT ((pquad_w_sub->subelement_type >= 0 && pquad_w_sub->subelement_type <= 15) || pquad_w_sub->subelement_type == -1);
```

Comment:
I have seen this 15 come up multiple times.

A general rule of programming says "No magic numbers" which means one should avoid having seemingly arbitrary numbers in their code, since it is prone to errors.

A solution to avoid magic numbers in C is to use macros.
If you define a macro name `T8_SUB_QUAD_MAXIMUM_TYPE` or similar and use this macro (or macro+1?) instead of 15, then a reader understands what it means.

The same holds for the 7 with ids

Reply:
Good advice, thanks. I replaced all subelement related constants by macros.